// @alchemy/aa-core/dist/esm/actions/smartAccount/buildUserOperationFromTx.js
import {} from "viem";
import { isBaseSmartAccountClient } from "../../client/isSmartAccountClient.js";
import { AccountNotFoundError } from "../../errors/account.js";
import { IncompatibleClientError } from "../../errors/client.js";
import { TransactionMissingToParamError } from "../../errors/transaction.js";
import { filterUndefined } from "../../utils/index.js";
import { buildUserOperation } from "./buildUserOperation.js";
export const buildUserOperationFromTx = async (client, args) => {
    const { account = client.account, ...request } = args;
    if (!account || typeof account === "string") {
        throw new AccountNotFoundError();
    }
    if (!request.to) {
        throw new TransactionMissingToParamError();
    }
    if (!isBaseSmartAccountClient(client)) {
        throw new IncompatibleClientError("BaseSmartAccountClient", "buildUserOperationFromTx");
    }
    const _overrides = {
        maxFeePerGas: request.maxFeePerGas ? request.maxFeePerGas : undefined,
        maxPriorityFeePerGas: request.maxPriorityFeePerGas
            ? request.maxPriorityFeePerGas
            : undefined,
    };
    filterUndefined(_overrides);
    return buildUserOperation(client, {
        uo: {
            target: request.to,
            data: request.data ?? "0x",
            value: request.value ? request.value : 0n,
        },
        overrides: _overrides,
        account: account,
    });
};
