// @alchemy/aa-core/dist/esm/middleware/actions.js
import { defaultFeeEstimator } from "./defaults/feeEstimator.js";
import { defaultGasEstimator } from "./defaults/gasEstimator.js";
import { defaultPaymasterAndData } from "./defaults/paymasterAndData.js";
import { noopMiddleware } from "./noopMiddleware.js";
export const middlewareActions = (overrides) => (client) => ({
    middleware: {
        customMiddleware: overrides.customMiddleware ?? noopMiddleware,
        dummyPaymasterAndData: overrides.paymasterAndData?.dummyPaymasterAndData
            ? async (struct) => ({
                ...struct,
                paymasterAndData: overrides.paymasterAndData.dummyPaymasterAndData(),
            })
            : defaultPaymasterAndData,
        feeEstimator: overrides.feeEstimator ?? defaultFeeEstimator(client),
        gasEstimator: overrides.gasEstimator ?? defaultGasEstimator(client),
        paymasterAndData: overrides.paymasterAndData?.paymasterAndData ?? defaultPaymasterAndData,
        userOperationSimulator: overrides.userOperationSimulator ?? noopMiddleware,
    },
});
