// @alchemy/aa-core/dist/cjs/utils/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultUserOperationFeeOptions = exports.minPriorityFeePerBidDefaults = exports.getDefaultSimpleAccountFactoryAddress = exports.getDefaultEntryPointAddress = void 0;
const index_js_1 = require("../chains/index.js");
const account_js_1 = require("../errors/account.js");
const entrypoint_js_1 = require("../errors/entrypoint.js");
const getDefaultEntryPointAddress = (chain) => {
    switch (chain.id) {
        case index_js_1.mainnet.id:
        case index_js_1.sepolia.id:
        case index_js_1.goerli.id:
        case index_js_1.polygon.id:
        case index_js_1.polygonMumbai.id:
        case index_js_1.optimism.id:
        case index_js_1.optimismGoerli.id:
        case index_js_1.optimismSepolia.id:
        case index_js_1.arbitrum.id:
        case index_js_1.arbitrumGoerli.id:
        case index_js_1.arbitrumSepolia.id:
        case index_js_1.base.id:
        case index_js_1.baseGoerli.id:
        case index_js_1.baseSepolia.id:
            return "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789";
    }
    throw new entrypoint_js_1.EntrypointNotFoundError(chain);
};
exports.getDefaultEntryPointAddress = getDefaultEntryPointAddress;
const getDefaultSimpleAccountFactoryAddress = (chain) => {
    switch (chain.id) {
        case index_js_1.mainnet.id:
        case index_js_1.polygon.id:
        case index_js_1.optimism.id:
        case index_js_1.optimismSepolia.id:
        case index_js_1.arbitrum.id:
        case index_js_1.base.id:
        case index_js_1.baseGoerli.id:
        case index_js_1.baseSepolia.id:
        case index_js_1.arbitrumSepolia.id:
            return "0x15Ba39375ee2Ab563E8873C8390be6f2E2F50232";
        case index_js_1.sepolia.id:
        case index_js_1.goerli.id:
        case index_js_1.polygonMumbai.id:
        case index_js_1.optimismGoerli.id:
        case index_js_1.arbitrumGoerli.id:
            return "0x9406Cc6185a346906296840746125a0E44976454";
    }
    throw new account_js_1.DefaultFactoryNotDefinedError("SimpleAccount", chain);
};
exports.getDefaultSimpleAccountFactoryAddress = getDefaultSimpleAccountFactoryAddress;
exports.minPriorityFeePerBidDefaults = new Map([
    [index_js_1.arbitrum.id, 10000000n],
    [index_js_1.arbitrumGoerli.id, 10000000n],
    [index_js_1.arbitrumSepolia.id, 10000000n],
]);
const getDefaultUserOperationFeeOptions = (chain) => {
    return {
        maxPriorityFeePerGas: {
            min: exports.minPriorityFeePerBidDefaults.get(chain.id) ?? 100000000n,
            percentage: 33,
        },
    };
};
exports.getDefaultUserOperationFeeOptions = getDefaultUserOperationFeeOptions;
