// @alchemy/aa-core/dist/cjs/utils/testUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTestClient = exports.createDummySmartContractAccount = void 0;
const viem_1 = require("viem");
const smartContractAccount_js_1 = require("../account/smartContractAccount.js");
const bundlerClient_js_1 = require("../client/bundlerClient.js");
const createDummySmartContractAccount = async (client, entrypoint) => {
    return (0, smartContractAccount_js_1.toSmartContractAccount)({
        source: "dummy",
        accountAddress: "0x1234567890123456789012345678901234567890",
        entrypointAddress: entrypoint,
        chain: client.chain,
        transport: (0, viem_1.custom)(client),
        signMessage: async () => "0xdeadbeef",
        signTypedData: async () => "0xdeadbeef",
        getAccountInitCode: async () => "0x1234567890123456789012345678901234567890deadbeef",
        encodeBatchExecute: async () => "0x",
        encodeExecute: async () => "0x",
        getDummySignature: () => "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c",
    });
};
exports.createDummySmartContractAccount = createDummySmartContractAccount;
const createTestClient = (chain) => {
    return (0, bundlerClient_js_1.createBundlerClientFromExisting)((0, viem_1.createPublicClient)({
        chain,
        transport: (0, viem_1.custom)({
            request: async () => { },
        }),
    }));
};
exports.createTestClient = createTestClient;
