// @alchemy/aa-core/dist/cjs/middleware/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.middlewareActions = void 0;
const feeEstimator_js_1 = require("./defaults/feeEstimator.js");
const gasEstimator_js_1 = require("./defaults/gasEstimator.js");
const paymasterAndData_js_1 = require("./defaults/paymasterAndData.js");
const noopMiddleware_js_1 = require("./noopMiddleware.js");
const middlewareActions = (overrides) => (client) => ({
    middleware: {
        customMiddleware: overrides.customMiddleware ?? noopMiddleware_js_1.noopMiddleware,
        dummyPaymasterAndData: overrides.paymasterAndData?.dummyPaymasterAndData
            ? async (struct) => ({
                ...struct,
                paymasterAndData: overrides.paymasterAndData.dummyPaymasterAndData(),
            })
            : paymasterAndData_js_1.defaultPaymasterAndData,
        feeEstimator: overrides.feeEstimator ?? (0, feeEstimator_js_1.defaultFeeEstimator)(client),
        gasEstimator: overrides.gasEstimator ?? (0, gasEstimator_js_1.defaultGasEstimator)(client),
        paymasterAndData: overrides.paymasterAndData?.paymasterAndData ?? paymasterAndData_js_1.defaultPaymasterAndData,
        userOperationSimulator: overrides.userOperationSimulator ?? noopMiddleware_js_1.noopMiddleware,
    },
});
exports.middlewareActions = middlewareActions;
