// @alchemy/aa-core/dist/cjs/actions/smartAccount/buildUserOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildUserOperation = void 0;
const isSmartAccountClient_js_1 = require("../../client/isSmartAccountClient.js");
const account_js_1 = require("../../errors/account.js");
const client_js_1 = require("../../errors/client.js");
const runMiddlewareStack_js_1 = require("./internal/runMiddlewareStack.js");
const buildUserOperation = async (client, args) => {
    const { account = client.account, overrides, uo } = args;
    if (!account) {
        throw new account_js_1.AccountNotFoundError();
    }
    if (!(0, isSmartAccountClient_js_1.isBaseSmartAccountClient)(client)) {
        throw new client_js_1.IncompatibleClientError("BaseSmartAccountClient", "buildUserOperation");
    }
    return (0, runMiddlewareStack_js_1._runMiddlewareStack)(client, {
        uo: {
            initCode: account.getInitCode(),
            sender: account.address,
            nonce: account.getNonce(),
            callData: Array.isArray(uo)
                ? account.encodeBatchExecute(uo)
                : typeof uo === "string"
                    ? uo
                    : account.encodeExecute(uo),
            signature: account.getDummySignature(),
        },
        overrides,
        account,
    });
};
exports.buildUserOperation = buildUserOperation;
