// @alchemy/aa-core/dist/cjs/actions/smartAccount/dropAndReplaceUserOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dropAndReplaceUserOperation = void 0;
const isSmartAccountClient_js_1 = require("../../client/isSmartAccountClient.js");
const account_js_1 = require("../../errors/account.js");
const client_js_1 = require("../../errors/client.js");
const index_js_1 = require("../../utils/index.js");
const runMiddlewareStack_js_1 = require("./internal/runMiddlewareStack.js");
const sendUserOperation_js_1 = require("./internal/sendUserOperation.js");
const dropAndReplaceUserOperation = async (client, args) => {
    const { account = client.account, uoToDrop, overrides } = args;
    if (!account) {
        throw new account_js_1.AccountNotFoundError();
    }
    if (!(0, isSmartAccountClient_js_1.isBaseSmartAccountClient)(client)) {
        throw new client_js_1.IncompatibleClientError("BaseSmartAccountClient", "dropAndReplaceUserOperation");
    }
    const uoToSubmit = {
        initCode: uoToDrop.initCode,
        sender: uoToDrop.sender,
        nonce: uoToDrop.nonce,
        callData: uoToDrop.callData,
        signature: uoToDrop.signature,
    };
    const { maxFeePerGas, maxPriorityFeePerGas } = await (0, runMiddlewareStack_js_1._runMiddlewareStack)(client, {
        uo: uoToSubmit,
        overrides,
        account,
    });
    const _overrides = {
        maxFeePerGas: (0, index_js_1.bigIntMax)(BigInt(maxFeePerGas ?? 0n), (0, index_js_1.bigIntPercent)(uoToDrop.maxFeePerGas, 110n)),
        maxPriorityFeePerGas: (0, index_js_1.bigIntMax)(BigInt(maxPriorityFeePerGas ?? 0n), (0, index_js_1.bigIntPercent)(uoToDrop.maxPriorityFeePerGas, 110n)),
    };
    const uoToSend = await (0, runMiddlewareStack_js_1._runMiddlewareStack)(client, {
        uo: uoToSubmit,
        overrides: _overrides,
        account,
    });
    return (0, sendUserOperation_js_1._sendUserOperation)(client, { uoStruct: uoToSend, account });
};
exports.dropAndReplaceUserOperation = dropAndReplaceUserOperation;
